arch/arc/core/userspace.S

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2017 Synopsys.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <offsets_short.h>
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <kernel_structs.h>
+#include <arch/cpu.h>
+#include <syscall.h>
+
+GTEXT(_arc_userspace_enter)
+GTEXT(_arc_do_syscall)
+GTEXT(_arch_is_user_context)
+
+
+/**
+ *
+ * User space entry function
+ *
+ * This function is the entry point to user mode from privileged execution.
+ * The conversion is one way, and threads which transition to user mode do
+ * not transition back later, unless they are doing system calls.
+ *
+ */
+SECTION_FUNC(TEXT, _arc_userspace_enter)
+	/*
+	 * In ARCv2, the U bit can only be set through exception return
+	 */
+	/* the end of user stack in r5 */
+	add r5, r4, r5
+	/* start of privilege stack */
+	add r2, r5, CONFIG_PRIVILEGED_STACK_SIZE
+	sub r5, r5, 16 /* skip r0, r1, r2, r3 */
+
+#ifdef CONFIG_INIT_STACKS
+	mov r0, 0xaaaaaaaa
+#else
+	mov r0, 0x0
+#endif
+_clear_user_stack:
+	st.ab r0, [r4, 4]
+	cmp r4, r5
+	jlt _clear_user_stack
+
+	lr r0, [_ARC_V2_STATUS32]
+	bset r0, r0, _ARC_V2_STATUS32_U_BIT
+
+	mov r1, _thread_entry_wrapper
+
+	/* fake exception return */
+	kflag _ARC_V2_STATUS32_AE
+
+	sr r0, [_ARC_V2_ERSTATUS]
+	sr r1, [_ARC_V2_ERET]
+
+	/* when exception returns from kernel to user, sp and _ARC_V2_USER_SP
+	 * will be switched
+	 */
+	sr r5, [_ARC_V2_USER_SP]
+	mov sp, r2
+
+	rtie
+
+/**
+ *
+ * Userspace system call function
+ *
+ * This function is used to do system calls from unprivileged code.  This
+ * function is responsible for the following:
+ * 1) Dispatching the system call
+ * 2) Restoring stack and calling back to the caller of the system call
+ *
+ */
+SECTION_FUNC(TEXT, _arc_do_syscall)
+	/* r0-r5: arg1-arg6, r6 is call id */
+	/* the call id is already checked in trap_s handler */
+	push_s blink
+	push_s r0
+
+	mov r0, _k_syscall_table
+	ld.as blink, [r0, r6]
+
+	pop_s r0
+
+	jl [blink]
+
+
+	pop_s blink
+
+
+	/* through fake exception return, go back to the caller */
+	kflag _ARC_V2_STATUS32_AE
+
+	/* the status and return addesss are saved in trap_s handler */
+	pop r6
+	sr r6, [_ARC_V2_ERSTATUS]
+	pop r6
+	sr r6, [_ARC_V2_ERET]
+	rtie
+
+SECTION_FUNC(TEXT, _arch_is_user_context)
```

Comment:
This is a very hot code path. Might be worth inlining using GCC extended asm. Then again it is a lot of instructions...

Reply:
The User mode status bit in Status32 is only readable in kernel mode, always returning 0 in user mode, so that can't be used for a simpler test.
However, there's another 'trick' that can be used.  If we allow user-mode sleep instructions by setting US to 1 (and require it’s not changed by applications), the test status32.US==1 is an efficient and robust way of detecting the current mode.
It enables a form of denial-of-service attack by putting the processor in sleep mode, but since interrupt level/mask can’t be set from user space that’s not worse than executing a loop without yielding.
And, a WDT could be used as additional protection.

Reply:
> since interrupt level/mask can’t be set from user space that’s not worse than executing a loop without yielding.

I concur.

Reply:
Well, mostly.
If a thread loops forever, it will starve any other threads of lower priority. But it could still get swapped out by another equal priority thread if time-slicing is turned on.
A higher priority thread could preempt it. But in this scenario this could only happen if some IRQ happens which wakes up that higher priority thread. Which would take the CPU out of sleep mode anyway.
So yeah, I think this is fine.

Reply:
Ruud's  suggestion is more efficient.  My concern is whether  user-mode sleep instructions bring other side effects, e.g., put other modules which require kernel privilege into sleep, e.g. mem controller, apex extension etc..